### DISCUSSION TITLE
簡單介紹opencv常用預處理圖像函式

### DISCUSSION THREAD

**Discussion by ghost** [1101-GS4719-C/Python-Community-GS4719-C#412] [General]
原圖(在下方程式經過調整，變為原圖0.5倍)：
![image](https://user-images.githubusercontent.com/100221228/156915173-f7fff1ac-c400-44f3-9c0c-b6236422ad42.png)
```
img = cv2.imread('kairyu.jpg')
img = cv2.resize(img, (0, 0), fx=0.5, fy=0.5)   # 調整大小倍率
img_gray = cv2.cvtColor(img, cv2.COLOR_BGR2GRAY)    # 灰階
```
第一行為讀取讀片， 
灰階結果：
![image](https://user-images.githubusercontent.com/100221228/156915119-4c3a04e1-ec99-437a-a43a-dabbc4a620bb.png)
```
img_GaussianBlur = cv2.GaussianBlur(img, (31, 31), 0)    
# 高斯模糊；圖片，高斯核(必定奇數)，標準差
```
高斯模糊結果：
![image](https://user-images.githubusercontent.com/100221228/156915250-c93e4edc-3f57-42c8-98a6-68bf6cbbea38.png)
```
img_Canny = cv2.Canny(img, 150, 200)
# 找邊緣，會在每個像素上打分數，與周圍差越多，分數越高；圖片，低於最低門檻則過濾，高於最高門檻全視作邊緣
```
找邊緣的結果：
![image](https://user-images.githubusercontent.com/100221228/156915305-7fb91606-7b45-47c5-b422-4ee08155e4a6.png)
```
kernal_dilate = np.ones((3, 3), np.uint8)  # create 預設維1的二維陣列
img_dilate = cv2.dilate(img_Canny, kernal_dilate, iterations=1)  # 膨脹；圖片，核，膨脹次數
```
膨脹剛剛找到邊緣的圖(第一個參數可見)：
![image](https://user-images.githubusercontent.com/100221228/156915348-5ab3586a-3d00-4034-8699-eb829a66e7ce.png)
```
kernal_erode = np.ones((5, 5), np.uint8)  # create 預設維1的二維陣列
img_erode = cv2.erode(img_dilate, kernal_erode, iterations=1)   
# 侵蝕，逆膨脹；圖片，核，侵蝕次數
```
侵蝕剛剛膨脹過邊緣的圖(第一個參數可見)：
![image](https://user-images.githubusercontent.com/100221228/156915377-04ef8777-79ad-48d3-a61b-6c7e77caa9b8.png)

為何膨脹還要侵蝕？不如說有時訓練一個人工智能的圖像辨識模型時，偶爾會利用到這過程，依照訓練之圖像數據而有不同的圖片預處理

把圖片show出來的函數如下
```
cv2.imshow('img', img)
cv2.imshow('img_gray', img_gray)
cv2.imshow('img_GaussianBlur', img_GaussianBlur)
cv2.imshow('img_Canny', img_Canny)
cv2.imshow('img_dilate', img_dilate)
cv2.imshow('img_erode', img_erode)
cv2.waitKey(0)
```

順便介紹一個不會教太難且生動風趣的程式教學youtuber
https://www.youtube.com/c/GrandmaCan%E6%88%91%E9%98%BF%E5%AC%A4%E9%83%BD%E6%9C%83

**Comment by pekoranoinu** [comment]
wow!好厲害啊!我也想學*-*

**Comment by chupochun** [comment]
原來還有改變圖像的程式函式讓我開了眼界

**Comment by ghost** [comment]
稍微推一下ww

**Comment by sean1325418** [comment]
腐蝕跟膨脹最簡單的目的就是去除雜訊，可以先用腐蝕去除雜訊，再用膨脹讓主要圖形變回來
